### ISSUE TITLE
Implement / fix TableField patch with tests

### ISSUE THREAD

**Issue by wesrowe** [department-of-veterans-affairs/va.gov-cms#10675]
## Description

In https://github.com/department-of-veterans-affairs/va.gov-cms/issues/10379 we found that the tablefield patch mostly worked, but a bug prevented the Editor from saving work.  

_Designs, per #10674._
_(updated 9/13 per Laura's comment below)_
![image](https://user-images.githubusercontent.com/4054752/190023719-69f7b5fd-c269-404a-b4e0-c711681d9e8a.png)

## Acceptance Criteria
- [ ] Check boxes to identify top row and/or first column as header appear as designed in #10674
- [ ] Saving the node with a tablefield paragraph succeeds without errors.
- [ ] Functional test(s) included in drupal.org contributed patch. Tests should validate the checkboxes exists and are functional.
- [ ] Code is contributed to Tablefield contrib module on drupal.org against [issue 3128030](https://www.drupal.org/project/tablefield/issues/3128030).
- [ ] After PR is tested and approved, @dsasser goes to https://www.drupal.org/project/tablefield/issues/3128030 and comments showing the patch was merged into our repo AND marks the issue as Reviewed and Tested by the Community.
- [ ] Requires design review (Val Rundle)

### CMS Team
Please check the team(s) that will do this work.

- [ ] `Program`
- [ ] `Platform CMS Team`
- [ ] `Sitewide Crew`
- [ ] `⭐️ Sitewide CMS`
- [x] `⭐️ Public Websites`
- [ ] `⭐️ Facilities`
- [ ] `⭐️ User support`

**Comment by laflannery** [comment]
@wesrowe I worked with Val to improve the inline guidance for both the Table title and the Table headers.  I was able to update the Figma file and also added a new screenshot below:
![image](https://user-images.githubusercontent.com/106678594/190000810-2001143b-aa10-4a76-8ca4-ce6b122aa0e9.png)

Let me know if there are any questions or you need anything else from me prior to implementation.

**Comment by EWashb** [comment]
For the horizontal/vertical checkboxes, are those just to add headers? The labels could be more descriptive.

**Comment by chri5tia** [comment]
@EWashb It just defines if the first row or first column is the header. The header data would still be entered or already entered into the table as per prior to any of these proposed changes.

**Comment by chri5tia** [comment]
Sanity check.
On the main branch, not the project branch where the patch was applied, I ran `composer install` and did database updates, yet I still see the row and column header fields. Whut.

![Screen Shot 2022-09-14 at 10 44 20 PM](https://user-images.githubusercontent.com/22764938/190323980-86932423-edd7-4155-933f-407891579a00.png)

**Comment by chri5tia** [comment]
I need a sanity check. After messing around a little in the db, I switched to the main branch where the patch is not applied, ran composer install and database updates, re-logged in to my local instance and I still see the row and column header fields. For fun I also re-imported configs with drush cim. When that didn't work, I pulled in the database from scratch! Must be something with sequalace. I clicked refresh tables, rebuild, etc. 

![Screen Shot 2022-09-14 at 11 24 44 PM](https://user-images.githubusercontent.com/22764938/190329992-a641743b-a5aa-475a-a4ec-e2495b2eefd8.png)

**Comment by dsasser** [comment]
> I need a sanity check. After messing around a little in the db, I switched to the main branch where the patch is not applied, ran composer install and database updates, re-logged in to my local instance and I still see the row and column header fields.

@chri5tia This is expected. Running updates is a one way thing, you can't undo an update by running updates again. To undo the tablefield patch changes, you need to reset your db `ddev pull va --skip-files`

**Comment by chri5tia** [comment]
Thanks for the sanity check. I am not sure what occurred after I did the db pull before, but it's good now.

```
MariaDB [db]> show columns in paragraph__field_table;
+---------------------+------------------+------+-----+---------+-------+
| Field               | Type             | Null | Key | Default | Extra |
+---------------------+------------------+------+-----+---------+-------+
| bundle              | varchar(128)     | NO   | MUL |         |       |
| deleted             | tinyint(4)       | NO   | PRI | 0       |       |
| entity_id           | int(10) unsigned | NO   | PRI | NULL    |       |
| revision_id         | int(10) unsigned | NO   | MUL | NULL    |       |
| langcode            | varchar(32)      | NO   | PRI |         |       |
| delta               | int(10) unsigned | NO   | PRI | NULL    |       |
| field_table_value   | longblob         | YES  |     | NULL    |       |
| field_table_format  | varchar(255)     | YES  |     | NULL    |       |
| field_table_caption | varchar(255)     | YES  |     | NULL    |       |
+---------------------+------------------+------+-----+---------+-------+
```

**Comment by chri5tia** [comment]
Removed the patch, (reverted the commit for composer.json which applied the patch), ran composer install, did a cache rebuild just because, and somehow I am still seeing this. I had also pulled in a fresh copy of the database again.

For some reason GitHub won't let me upload the screenshot. 

Starting on line 55 of tablefield.shema.yml (the one in tablefield/config/schema/

```
field.formatter.settings.tablefield:
  type: mapping
  label: 'Tablefield field widget settings'
  mapping:
    row_header:
      type: boolean
      label: 'Display first row as a table header'
    column_header:
      type: boolean
      label: 'Display first column as a table header'
```

**Comment by chri5tia** [comment]
Fighting a lot with this. Getting a lot of conflicts trying to rebase even after doing a `git reset --hard head` to remove all of my work and commits. Need to create a new project branch and close the existing draft PR. 

Wondering if the tablefield module still holds enough value instead of other methods for table-style content creation:
a) paragraph template for custom paragraph table type
b) wysiwyg table insert for rich text
c) https://www.drupal.org/project/paragraphs_table

In philosophy, there is a lot of practical value in not trying to reinvent the wheel or change things up to drastically, improving on existing software and methods, but at some point we may need to evaluate how much we've evolved away from tablefield continuing to be viable, considering all the patches we've already applied to it.

Will definitely need to rope in more minds to move this forward.

Other observations about the patch and/or the tablefield module:
- Boolean checkbox isn't in the list of allowed field types
- D8 version of tablefield doesn't support search api indexing

Here is my draft patch for collab:
https://github.com/chri5tia/patchwork/blob/patches/tablefield.patch

**Comment by wesrowe** [comment]
@laflannery, does 508 compliance require every table to have at least one header (whether it's at the top or left)? Val raised the question, as it would affect whether the new checkboxes are required. It would also affect whether we would continue to have the FE default to top-row headers.

**Comment by chri5tia** [comment]
@dsasser Putting this here for now while we digest things:
https://github.com/chri5tia/patchwork/blob/main/patches/tablefield1.patch
These are the changes to move the form elements to the widget.

**Comment by chri5tia** [comment]
As @dsasser and I work through troubleshooting, fixing and building out this feature, I am keeping track of things we've tried that don't necessarily relate to each other as separate scrap patches:
https://github.com/chri5tia/patchwork/tree/main/patches

**Comment by laflannery** [comment]
@wesrowe Data tables should always have at least one header specified. Layout/presentation tables (tables used only to organize content into columns) do not need these headers.  However, I do not believe this table widget is intended for simply layout purposes so I do not believe that case applies here.  That case would also require additional markup if it does in fact apply. I had noticed that the current editor UI had the note about leaving the row blank, but I did not see this in the Figma screenshot so I thought that functionality/guidance it was being removed with this new patch.

Regarding defaults, during one meeting it was discussed how to handle defaulting the headers and I believe we had decided not to default within the UI but instead to create a FE template for top-row in order to assist with the QA/fixing of the tables.  That way once all existing table were updated correctly, this default template could be removed and all new tables being created would have to select a header choice or it would not save properly.

**Comment by jilladams** [comment]
Per discussion with @wesrowe : this issue is blocked by a deeper design think & content audit in #10793. More context on the why forthcoming. Meantime: I've asked Daniel to update ticket so we can pause work for now and make it easy to resume later.

**Comment by dsasser** [comment]
# Developer notes

## Summarized development timeline
- We tested the tablefield [patch](https://www.drupal.org/project/tablefield/issues/3128030#comment-13830672) from the [existing d.o issue](https://www.drupal.org/project/tablefield/issues/3128030), and found we were receiving integrity constraint violations when saving.
- After a lot of debugging, we found the cause was related to some custom code in va_gov_backend::_va_gov_backend_apply_filter_to_tablefield_cells() which is called during an entity_presave hook.
- After fixing this, we re-tested the tablefield patch, and found that it _still_ didn't persist data into the database.
- We also found that the update hook in this patch was not accounting for field storage definition updates. We resolved this with our own update hook implementation.
- We found the issue in the tablefield patch and created [a new patch against the dev version of the module,](https://www.drupal.org/project/tablefield/issues/3128030#comment-14701623) as well as a [re-rolled patch](https://www.drupal.org/files/issues/2022-09-22/8.x-2.3_tablefield-header-orientation-3128030-19.patch) for the current released version: 2.3.

## Where we are today
We have a [va.gov-cms PR](https://github.com/department-of-veterans-affairs/va.gov-cms/pull/10809) that includes: adding the updated tablefield patch to composer, the update hook to solve for field storage definition updates, and the bugfix in va_gov_backend. This PR was, at the time, a complete solution to this issue. However, it was not long after the PR was opened that we were notified a design review process was going to kick off--and thus a pause on this issue.

## For when we pick this back up
The updated tablefield patch is an iteration of the original community provided patch, but the implementation approach could use some additional review. The community appears to be bisected on the path for the D8 version of tablefield, as it pertains to adding display options to the table field. One approach, the one we took, was to include the heading orientation as a pair of new boolean fields, presented as checkboxes on the table edit form. [The other approach](https://www.drupal.org/project/tablefield/issues/3019613#comment-13360497) is one worth looking into however, as it utilizes a single database field to represent the heading choice. A single field means a smoother UX (fewer input elements) and less custom code (we wouldn't need to figure out how to "require" a checkbox without making it always checked--an implied requirement from the original design).

Further, consider [this comment](https://www.drupal.org/project/tablefield/issues/3019613#comment-13374685) from the module maintainer, which suggests a sub-module for moving to entity-level override-able storage options for all tablefield formatters.

**Comment by jilladams** [comment]
Note: Swirt has specifically suggested (somewhere in Slack?) that we not go the submodule route.

**Comment by jilladams** [comment]
https://github.com/department-of-veterans-affairs/va.gov-cms/issues/10800#issuecomment-1318650289
Closing
